django/db/migrations/recorder.py

```diff
@@ -32,8 +20,24 @@ class Meta:
         def __str__(self):
             return "Migration %s for %s" % (self.name, self.app)
 
+    return Migration
+
+
+class MigrationRecorder:
+    """
+    Deal with storing migration records in the database.
+
+    Because this table is actually itself used for dealing with model
+    creation, it's the one thing we can't do normally via migrations.
+    We manually handle table creation/schema updating (using schema backend)
+    and then have a floating model to do queries with.
+
+    If a migration is unapplied its row is removed from the table. Having
+    a row in the table always means a migration is applied.
+    """
     def __init__(self, connection):
         self.connection = connection
+        self.Migration = migration()
```

Comment:
> Have we explored making MigrationRecorder.Migration to be lazily created on the first class access instead?

Will this have the `Migration` class created each time `MigrationRecorder` is instantiated? Any issues with that? Any preference for making it a static attribute on the class, so it's only created once?

Reply:
@carltongibson No class definitions are only created once when the interpreter reaches in the function `migration`. `self.Migration = migration()` will surely set Migration class again into `self.Migration`, but python interpreter only creates the class definition the first time it encounters that class. 
Moreover, making it static won't fix the issue. If we make it static importing this file would trigger the `migration()` function and we would be back where we started.

Reply:
@carltongibson What do you think if we make it as an attribute?
like
```python
    @property
    def Migration(self):
        class Migration(models.Model):
            app = models.CharField(max_length=255)
            name = models.CharField(max_length=255)
            applied = models.DateTimeField(default=now)

            class Meta:
                apps = Apps()
                app_label = "migrations"
                db_table = "django_migrations"

            def __str__(self):
                return "Migration %s for %s" % (self.name, self.app)

        return Migration
```

Reply:
Is that correct?

```
>>> def a():
...     class A:
...         pass
...         
...     return A
...
>>> a1 = a()
>>> a2 = a()
>>> a1
__main__.a.<locals>.A
>>> a2
__main__.a.<locals>.A
>>> a1 is a2
False
>>> a1 == a2
False
```

Whereas if `Migration` is a `@property` that lazily creates `Migration`, storing the reference on the class, it'll always be the same. 

I'm not 100% sure it'd raise any problems but, `ModelBase.__new__()` is a little bit hairy, and dynamic model creation isn't really supported, I think for that reason. Hence my query.

Reply:
@nasirhjafri Our comments crossed. 

Take this class: 

```
class T: 
    
    _B = None
     
    @property 
    def a(self): 
        class A: 
            pass 
    
        return A
    
    @property
    def b(self):
        if self.__class__._B is not None: 
            class B:
                pass
            
            self.__class__._B = B
        return self.__class__.B
```

Behaviour is thus: 

```
>>> T
__main__.T
>>> t = T()
>>> a1 = t.a
>>> a2 = t.a
>>> a1 is a2
False
>>> b1 = T.b
>>> b2 = T.b
>>> b1 is b2
True
```

Your example is like `a()`, whereas `b()` is more like what I took Simon's initial comment to mean. I don't know if it's the best idiom, or if it's 100% necessary, but that's the sort of thing I have in mind.

Reply:
Clearly using `__class__` isn't what I mean here.

Reply:
I think @carltongibson's suggestion is a good one. The `Migraiton` class should only ever be created once. Using the `classproperty` decorartor in `django/utils/decorators.py` may make the code simpler.

Reply:
Sorry for the double post, but just to be explicit, something like:

```py
@classproperty
def migration(cls):
    if cls._migration_class is None:
        class Migration:
            ...

        cls._migration_class = Migration
    return cls._migration_class
```

Reply:
Thanks @jdufresne. I knew there has to be a nicer way... 🙂

Reply:
Hi @carltongibson, I've improvised my fix. Can you please review it?
Thanks